synapse/storage/devices.py

```diff
@@ -127,16 +129,21 @@ def get_devices_by_remote(self, destination, from_stream_id, limit):
         # (user_id, device_id) entries into a map, with the value being
         # the max stream_id across each set of duplicate entries
         #
-        # maps (user_id, device_id) -> stream_id
+        # maps (user_id, device_id) -> (stream_id, context)
         # as long as their stream_id does not match that of the last row
+        # where context is any metadata about the message's context such as
+        # opentracing data
         query_map = {}
         for update in updates:
             if stream_id_cutoff is not None and update[2] >= stream_id_cutoff:
                 # Stop processing updates
                 break
 
             key = (update[0], update[1])
-            query_map[key] = max(query_map.get(key, 0), update[2])
+            context = update[3]
+            stream_id = max(query_map.get(key, (0, None))[0], update[2])
+
+            query_map[key] = (stream_id, context)
```

Comment:
I guess the choice of which `context` to use if there are multiple candidates is pretty arbitrary, but have you had any thoughts on it?

Reply:
The context is related the the update which is acted on, in this case it's always the latest. We could link all the contexts but Jaeger has poor support for this and I don't think it will tell us any more than that earlier updates were ignored. I think it's fine this way.

Reply:
Although it knowing that they were explicitly ignored is useful. I could start a child span here with the purpose of stating that the update had no effect.

Reply:
The context related to the update which is ultimately acted on is the most important since the updates can differ. It could be useful to know that the previous updates were ignored, however that would already be pretty clear.

Reply:
So the context of the chosen update should be in the query map.